SYNTH-875 : Added a set of AppDynamics capabilities

shashwatsrivastava94 opened the pull request:
Added a set of new capabilities which will allow us to run the webdriver server
as a long running process.
The capabilities look like this right now -
appdynamicsCapability = {"commandWebhook" : "something", "testId" : "1234", "outputDir" : "path to outputDir"}

Commit:
SYNTH-875 : Added a set of AppDynamics capabilities

Added a set of new capabilities which will allow us to run the webdriver server
as a long running process.
The capabilities look like this right now -
appdynamicsCapability = {"commandWebhook" : "something", "testId" : "1234", "outputDir" : "path to outputDir"}

renajohn commented on java/client/src/org/openqa/selenium/AppdynamicsCapabilities.java:
```diff
@@ -0,0 +1,97 @@
+  /**
+   * This capability sets the K9 output directory.
+   */
+  public static String K9_OUTPUT_DIR = "outputDir";
```
Let's drop K9_ prefix. We might be using this output dir for other capabilities

renajohn commented on java/client/src/org/openqa/selenium/AppdynamicsCapabilities.java:
```diff
@@ -0,0 +1,97 @@
+  /**
+   * This capability sets the K9 test id
+   */
+  public static String K9_TEST_ID = "testId";
```
Same here

renajohn commented on java/client/src/org/openqa/selenium/AppdynamicsCapabilities.java:
```diff
@@ -0,0 +1,97 @@
+/**
+ * Created by shashwat.srivastava on 5/30/17.
+ */
+public class AppdynamicsCapabilities {
```
Let's call it `AppdynamicsCapability` since we decided to go for a single capability

renajohn commented on java/client/src/org/openqa/selenium/AppdynamicsCapabilities.java:
```diff
@@ -0,0 +1,97 @@
+  public static final String APPDYNAMICS_CAPABILITIES = "appdynamicsCapabilities";
+
+  public AppdynamicsCapabilities(Map<String, String> map) {
+    if (map.containsKey(K9_OUTPUT_DIR) && map.get(K9_OUTPUT_DIR) != null) {
```
Should we test for empty string? I don't remember if Selenium as Strings.isNullOrEmpty(), if it has it, let's use it.

renajohn commented on java/client/src/org/openqa/selenium/AppdynamicsCapabilities.java:
```diff
@@ -0,0 +1,97 @@
+    if (map.containsKey(K9_OUTPUT_DIR) && map.get(K9_OUTPUT_DIR) != null) {
+      this.outputDir = map.get(K9_OUTPUT_DIR);
+    }
+    if (map.containsKey(K9_TEST_ID) && map.get(K9_TEST_ID) != null) {
```
Same here

renajohn commented on java/client/src/org/openqa/selenium/AppdynamicsCapabilities.java:
```diff
@@ -0,0 +1,97 @@
+      }
+    }
+    return appdynamicsCapabilities;
+  }
```
Drop K9 in all variables

renajohn commented on java/client/src/org/openqa/selenium/chrome/ChromeDriver.java:
```diff
@@ -171,7 +172,8 @@ public ChromeDriver(ChromeDriverService service, ChromeOptions options) {
   public ChromeDriver(ChromeDriverService service, Capabilities capabilities) {
-    super(new ChromeDriverCommandExecutor(service), capabilities);
+    super(new ChromeDriverCommandExecutor(service.withEnvironment(
+      AppdynamicsCapabilities.extractFrom(capabilities).getK9Environment())), capabilities);
```
Chrome doesn't need K9.

Commit:
SYNTH-875 : Addressed Renaults CR comments

Commit:
SYNTH-875 : No longer need to use the save screenshots environment variable

The selenium server now uses the output directory being sent as a capability
to create the screenshots folder.

renajohn commented on java/client/src/org/openqa/selenium/AppdynamicsCapability.java:
```diff
@@ -0,0 +1,98 @@
+   */
+  public static String OUTPUT_DIR = "outputDir";
+  /**
+   * This capability sets the K9 test id
```
Remove K9 from the comment

renajohn commented on java/client/src/org/openqa/selenium/AppdynamicsCapability.java:
```diff
@@ -0,0 +1,98 @@
+   * This capability will make it possible to set environment variables for
+   * the K9 process
+   */
+  public static final String APPDYNAMICS_CAPABILITIES = "appdynamicsCapabilities";
```
Should be `capability`

renajohn commented on java/server/src/org/openqa/selenium/remote/server/handler/CaptureScreenshot.java:
```diff
@@ -38,7 +39,9 @@
-    String path = System.getProperty("webdriver.saveScreenshots");
+    AppdynamicsCapability appdynamicsCapability = AppdynamicsCapability.extractFrom(session.getCapabilities());
+
+    String path = appdynamicsCapability.getOutputDir() + "\\screenshots";
```
Yeah! 👍  :)

Commit:
SYNTH-875 : Addressed Renaults comments
